core/src/kura.rs

```diff
@@ -4,258 +4,487 @@
-            .read_all()
-            .await?
-            .try_collect::<Vec<_>>()
-            .await?;
-        self.merkle_tree = blocks
-            .iter()
-            .map(VersionedCommittedBlock::hash)
-            .collect::<MerkleTree<_>>();
-        Ok(blocks)
+impl BlockStoreTrait for StdFileBlockStore {
+    #[allow(
+        clippy::needless_range_loop,
+        clippy::unwrap_used,
+        clippy::unwrap_in_result
+    )]
+    fn read_block_indices(
+        &self,
+        start_block_height: u64,
+        dest_buffer: &mut [(u64, u64)],
+    ) -> Result<()> {
+        let mut index_file = std::fs::OpenOptions::new()
+            .read(true)
+            .open(self.path_to_blockchain.join("blocks.index"))?;
+        let start_location = start_block_height * 2 * 8;
+        let block_count = dest_buffer.len() as u64;
+
+        if start_location + (2 * 8) * block_count > index_file.metadata()?.len() {
+            return Err(Error::OutOfBoundsBlockRead(start_block_height, block_count));
+        }
+        index_file.seek(SeekFrom::Start(start_location))?;
+        // (start, length), (start,length) ...
+        for i in 0..block_count {
+            let index: usize = i.try_into().unwrap();
+            dest_buffer[index] = (
+                index_file.read_u64::<LittleEndian>()?,
+                index_file.read_u64::<LittleEndian>()?,
+            );
+        }
+        Ok(())
     }
 
-    /// Methods consumes new validated block and atomically stores and caches it.
-    #[iroha_futures::telemetry_future]
-    #[log("INFO", skip(self, block))]
-    pub async fn store(
+    fn write_block_index(&mut self, block_height: u64, start: u64, length: u64) -> Result<()> {
+        let mut index_file = std::fs::OpenOptions::new()
+            .write(true)
+            .create(true)
+            .open(self.path_to_blockchain.join("blocks.index"))?;
+        let start_location = block_height * 2 * 8;
+        if start_location + 2 * 8 > index_file.metadata()?.len() {
+            index_file.set_len(start_location + 2 * 8)?;
+        }
+        index_file.seek(SeekFrom::Start(start_location))?;
+        // block0       | block1
+        // start, length| start, length  ... et cetera.
+        index_file.write_u64::<LittleEndian>(start)?;
+        index_file.write_u64::<LittleEndian>(length)?;
+        Ok(())
+    }
+
+    #[allow(clippy::integer_division)]
+    fn read_index_count(&self) -> Result<u64> {
+        let index_file = std::fs::OpenOptions::new()
+            .read(true)
+            .open(self.path_to_blockchain.join("blocks.index"))?;
+        Ok(index_file.metadata()?.len() / (2 * std::mem::size_of::<u64>() as u64)) // Each entry is 16 bytes.
+    }
+
+    fn write_index_count(&mut self, new_count: u64) -> Result<()> {
+        let index_file = std::fs::OpenOptions::new()
+            .write(true)
+            .open(self.path_to_blockchain.join("blocks.index"))?;
+        let new_byte_size = new_count * 2 * 8;
```

Comment:
can we extract these magic numbers into some conversion function?

Reply:
We could, but that would make it less readable in my opinion. It is true that we write this a couple of times. But the size of a function call vs this statement is the same. And with it this way I the reader can actually see what the value is without having to look it up.

Reply:
seems like a difference in opinions. For me, as someone who hasn't been deep into `kura`, this is less readable than a function with a good descriptive name. Can we at least use constants instead of magic numbers?

Reply:
yes